### ISSUE TITLE
Extremely slow with hdf5 dataset, consuming all memory

### ISSUE THREAD

**Issue by liueryun** [tflearn/tflearn#54]
My problem:
----------------------------------------------------------------
I have a huge hdf5 dataset stored in 'data.h5'. The shape of X is (2000000,40,40,3). The following code consumes all my memory. It seems the program loads all the data into memory. What is the problem?

Thank you.

My code:
----------------------------------------------------------------
 # Load hdf5 dataset
 h5f = h5py.File('data/data.h5', 'r')
 X = h5f['fre_ori_train']
 Y = h5f['targets_train']
 X_test = h5f['fre_ori_val'][:10000,:,:,:]
 Y_test = h5f['targets_val'][:10000,:]
 
 print(X.shape)
 print(Y.shape)
  
 # Build network
 network = input_data(shape=[None, 40, 40, 3], dtype=tf.float32)
 network = conv_2d(network, 32, 3, activation='relu')
 network = max_pool_2d(network, 2)
 network = conv_2d(network, 64, 3, activation='relu')
 network = conv_2d(network, 64, 3, activation='relu')
 network = max_pool_2d(network, 2)
 network = fully_connected(network, 512, activation='relu')
 network = dropout(network, 0.5)
 network = fully_connected(network, 5, activation='linear')
 network = regression(network, optimizer='adam',
                      loss='mean_square',
                      learning_rate=0.001)
 
 # Training
 model = tflearn.DNN(network, tensorboard_verbose=0)
 model.fit(X, Y, n_epoch=50, shuffle=False, validation_set=(X_test, Y_test),
           show_metric=True, batch_size=32, run_id='fre_ori_cnn')
 
 h5f.close()
 ----------------------------------------------------------------

**Comment by aymericdamien** [comment]
Hi,
About the memory issue, I think that this code:
```
X_test = h5f['fre_ori_val'][:10000,:,:,:]
Y_test = h5f['targets_val'][:10000,:]
```
actually returns a numpy array, so it loads all data in memory. Can you try to remove it and remove the validation_set in the model definition and see if it fix the issue?

Maybe we need to further investigate memory and speed impact of large hdf5 dataset.

**Comment by liueryun** [comment]
Thanks @aymericdamien !

I tried to comment out these two lines and got the following errors.
`Exception: R2 metric requires Inputs and Targets to have same shape.`

The shape of Y is (2000000,5). Should I reshape it?

BTW, the 'data.h5' file is about 50G. My system is Mac OS Yosemite, 8G RAM.

**Comment by aymericdamien** [comment]
Did you specify to use the R2 metric in the regression layer? If yes, then it requires X and Y data to have same shape. If not, it may be a bug, the default metric should be top-1 accuracy.

**Comment by liueryun** [comment]
Yes, I used R2 metric. The shapes of X and Y are indeed different. The output Y is a 5 dimentional vector. What I really want to do in the regression layer is to comput the Euclidean distance between predicted values and ground truth as the metric of accuracy. What is the right name of  metric? I really appreciate your help.

**Comment by aymericdamien** [comment]
When you are using the mean square error, it is already computing a distance between X and Y. If you want to add a new metric, you can add a custom class based on this template: https://github.com/tflearn/tflearn/blob/master/tflearn/metrics.py#L128

Basically, predictions are predicted values f(X), targets is Y and inputs is X. You can build any custom new metric and pass it as argument of `regression` layer.
In your case, should be something like:
```
class EuclidianDistance(Metric):

    def __init__(self, name=None):
        super(EuclidianDistance, self).__init__(name)
        self.name = "EuclidianDistance" if not name else name

    def build(self, predictions, targets, inputs):
        """ Build euclidian distance tensor. """
        self.built = True
        self.tensor = YOUR_DISTANCE_OP(predictions, targets, inputs)
        # Add a special name to that tensor, to be used by monitors
        self.tensor.m_name = self.name
```

`YOUR_DISTANCE_OP` is the custom operation that will compute the score.
